cpp/src/arrow/adapters/orc/jni/jni_wrapper.cpp

```
@@ -0,0 +1,243 @@
+  return (*reader)->NumberOfStripes();
+}
+
+JNIEXPORT jobject JNICALL
+Java_org_apache_arrow_adapter_orc_OrcReaderJniWrapper_nextStripeReader(JNIEnv* env,
+                                                                       jobject this_obj,
+                                                                       jlong batch_size) {
+  auto reader = GetNativeReader(env, this_obj);
+  auto stripe_reader = new std::shared_ptr<RecordBatchReader>();
+  auto status = (*reader)->NextStripeReader(batch_size, stripe_reader);
+
+  jobject ret = nullptr;
+  if (!status.ok()) {
+    delete stripe_reader;
+    return ret;
+  }
+
+  jclass cls = env->FindClass("/org/apache/arrow/adapter/orc/OrcStripeReaderJniWrapper");
+  ret = env->AllocObject(cls);
+
+  env->SetLongField(ret, GetFieldId(env, cls, "nativeStripeReaderAddress"),
+                    reinterpret_cast<long>(stripe_reader));
+
+  return ret;
+}
+
+JNIEXPORT jbyteArray JNICALL
+Java_org_apache_arrow_adapter_orc_OrcStripeReaderJniWrapper_getSchema(JNIEnv* env,
+                                                                      jobject this_obj) {
+  auto stripe_reader = GetStripeReader(env, this_obj);
+  auto schema = (*stripe_reader)->schema();
+
+  std::shared_ptr<arrow::Buffer> out;
+  auto status = arrow::ipc::SerializeSchema(*schema, arrow::default_memory_pool(), &out);
+  if (!status.ok()) {
+    return nullptr;
+  }
+
+  jbyteArray ret = env->NewByteArray(out->size());
+  memcpy(env->GetByteArrayElements(ret, nullptr), out->data(), out->size());
+  return ret;
+}
+
+JNIEXPORT jobject JNICALL
+Java_org_apache_arrow_adapter_orc_OrcStripeReaderJniWrapper_next(JNIEnv* env,
+                                                                 jobject this_obj) {
+  auto stripe_reader = GetStripeReader(env, this_obj);
+  auto record_batch = new std::shared_ptr<arrow::RecordBatch>();
+  auto status = (*stripe_reader)->ReadNext(record_batch);
+  if (!status.ok()) {
+    delete record_batch;
+    return nullptr;
```

should we return an error instead..how does the client know why the read failed..